document-generation/igce-document.ts

```diff
@@ -0,0 +1,270 @@
+import { INTERNAL_SERVER_ERROR } from "../utils/errors";
+import { ApiBase64SuccessResponse, SuccessStatusCode } from "../utils/response";
+import { convertPeriodToMonths } from "./utils/utils";
+
+export async function generateIGCEDocument(
+  templatePath: string,
+  payload: IndependentGovernmentCostEstimate
+): Promise<ApiBase64SuccessResponse> {
+  if (!payload.periodsEstimate && !payload.fundingDocument && !payload.surgeCapabilities) {
+    return INTERNAL_SERVER_ERROR;
+  }
+
+  // sort the base period and option periods
+  const basePeriodLineItems = payload.periodsEstimate.filter((periodEstimate: IPeriodEstimate) => {
+    return periodEstimate.period.periodType === PeriodType.BASE;
+  });
+  const optionPeriodsLineItems = payload.periodsEstimate.filter((periodEstimate: IPeriodEstimate) => {
+    return periodEstimate.period.periodType === PeriodType.OPTION;
+  });
+
+  const workbook = new exceljs.Workbook();
+  await workbook.xlsx.readFile(templatePath);
+
+  // variables used within the sheets
+  let fundingDocumentNumber: string;
+  if (!payload.fundingDocument.gtcNumber && !payload.fundingDocument.orderNumber) {
+    fundingDocumentNumber = `MIPR Number: ${payload.fundingDocument.miprNumber}`;
+  } else {
+    const { orderNumber, gtcNumber } = payload.fundingDocument;
+    fundingDocumentNumber = `Order Number: ${orderNumber} and GT&C Number: ${gtcNumber}`;
+  }
+
+  const summarySheet = workbook.getWorksheet("Summary");
+  const allPeriodsIdiqClins: string[] = [];
+  const summarySheetColumnsToExtend: string[] = []; // keep track of columns to extend
+  let summarySheetIdiqClinCounter = 0; // keep count of unique idiq clins of each period
+
+  // function that populates each period sheet
+  function populatePeriodLineItems(estimate: IPeriodEstimate): void {
+    const { optionOrder, periodUnit, periodUnitCount, periodType } = estimate.period;
+    let periodSheet: Worksheet;
+    const periodLineItems = estimate.periodLineItems;
+    const uniqueIdiqClins: string[] = [];
+    const periodSheetTopHalfInitialStartRow = 8;
+    const periodSheetLowerHalfInitialStartRow = 28;
+
+    if (periodType === PeriodType.BASE) {
+      periodSheet = workbook.getWorksheet("Base Period");
+    } else {
+      periodSheet = workbook.getWorksheet(`Option Period ${optionOrder}`);
+    }
+    const pop = `${periodUnitCount} ${periodUnit[0] + periodUnit.slice(1).toLowerCase()}(s)`;
+    periodSheet.getCell("C2").value = pop;
+    periodSheet.getCell("C3").value = fundingDocumentNumber;
+
+    // add each line item to the period sheet (top half)
+    const numberOfItems = periodLineItems.length;
+    const lineItemRows = periodSheet.getRows(periodSheetTopHalfInitialStartRow, numberOfItems);
+    lineItemRows?.forEach((row, index) => {
+      if (!uniqueIdiqClins.includes(periodLineItems[index].idiqClin)) {
+        uniqueIdiqClins.push(periodLineItems[index].idiqClin);
+      }
```

Comment:
This may be a little cleaner? Not sure.
```suggestion
    const uniqueIdiqClins = Array.from(
      new Set(periodLineItems.map((lineItem) => lineItem.idiqClin))
    );
    lineItemRows?.forEach((row, index) => {
```

Reply:
yeah, I think so.